### PR TITLE
JDK-8293472: Incorrect container resource limit detection if manual cgroup fs mounts present

### PR DESCRIPTION
**Pull request by casparcwang**
Similar to [JDK-8253435](https://bugs.openjdk.org/browse/JDK-8253435), when setting the mount path of cgroup controller "memory/cpu/cpuacct/pids", it should also discard duplicate items.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293472](https://bugs.openjdk.org/browse/JDK-8293472): Incorrect container resource limit detection if manual cgroup fs mounts present

### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [c4109d65](https://git.openjdk.org/jdk/pull/10193/files/c4109d65e95daae20aa178658495bd69f1688cd1)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10193/head:pull/10193` \
`$ git checkout pull/10193`

Update a local copy of the PR: \
`$ git checkout pull/10193` \
`$ git pull https://git.openjdk.org/jdk pull/10193/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10193`

View PR using the GUI difftool: \
`$ git pr show -t 10193`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10193.diff">https://git.openjdk.org/jdk/pull/10193.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (jerboaa) [inline src/hotspot/os/linux/cgroupSubsystem_linux.cpp:136]: Please demote this to `log_debug` over a `log_warning`.

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (jerboaa) [inline src/hotspot/os/linux/cgroupSubsystem_linux.cpp:141]: Please demote this to `log_debug` over `log_warning`.

**Review by jerboaa — CHANGES_REQUESTED** [review]
Looks mostly good. https://bugs.openjdk.org/browse/JDK-8270087 seems to be an early sighting of this. It mentions a reproducer, manually mount `/sys/fs/cgroup` from the host into the container. Please write a container test asserting that no warnings get produced. Perhaps use `test/hotspot/jtreg/containers/docker/DockerBasicTest.java` and write a new test method `javaVersionWithCgMounts` or some such, which verifies no warning shows up on stdout.

**Comment by jerboaa** [comment]
My testing on cgv2 and cgv1 indicates that we might have a container resource limit detection problem depending on the order of `mountinfo` entries are observed at runtime. Please add memory and cpu limit container detection tests which do `-v /sys/fs/cgroup:/cgroups-in:ro` as a volume mount in addition. Thanks!

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (casparcwang) [reply src/hotspot/os/linux/cgroupSubsystem_linux.cpp]: done

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (casparcwang) [reply src/hotspot/os/linux/cgroupSubsystem_linux.cpp]: done

**Comment by casparcwang** [comment]
> Looks mostly good. https://bugs.openjdk.org/browse/JDK-8270087 seems to be an early sighting of this. It mentions a reproducer, manually mount `/sys/fs/cgroup` from the host into the container. Please write a container test asserting that no warnings get produced. Perhaps use `test/hotspot/jtreg/containers/docker/DockerBasicTest.java` and write a new test method `javaVersionWithCgMounts` or some such, which verifies no warning shows up on stdout.

Thank you very much for the reviewing. I have not noticed JDK-8270087 before, sorry about filing a duplicate jdk bug.

> My testing on cgv2 and cgv1 indicates that we might have a container resource limit detection problem depending on the order of `mountinfo` entries are observed at runtime. Please add memory and cpu limit container detection tests which do `-v /sys/fs/cgroup:/cgroups-in:ro` as a volume mount in addition. Thanks!

The `container resource limit detection problem` is also fixed when there are multiple cgroup fs mount entries, it is caused by multiple setting of root mount path in function `CgroupSubsystemFactory::determine_type`.

File: test/hotspot/jtreg/containers/docker/DockerBasicTest.java
Comment (jerboaa) [inline test/hotspot/jtreg/containers/docker/DockerBasicTest.java:102]: Those are all cgroups v1 specific paths, so the test will fail when run on cgroups v2.

File: test/hotspot/jtreg/containers/docker/DockerBasicTest.java
Comment (jerboaa) [inline test/hotspot/jtreg/containers/docker/DockerBasicTest.java:91]: I'd suggest, to first add a test which verifies that without `-Xlog:os+container=trace` option, `java -version` on such a container doesn't print **any** `os+container` logs (it used to print a warning in the` -v /sys/fs/cgroup:/cgroups-in:ro` case). That's all this test should verify.

**Review by jerboaa — CHANGES_REQUESTED** [review]
We definitely need regression tests for memory/cpu limits with those additional cgroup mounts. 
Please take a look at `test/hotspot/jtreg/containers/docker/TestCPUAwareness.java` and `test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java`

For cgroups v2 the issues isn't fixed. The paths for cg v2 are being set here:
https://github.com/openjdk/jdk/blob/98da03af50e2372817a7b5e381eea5ee6f2cb919/src/hotspot/os/linux/cgroupSubsystem_linux.cpp#L303..L313.

Example of running (with your patch on a cg v2 system):

```
$ sudo podman run --rm -ti --memory=300M --memory-swap=300M -v /sys/fs/cgroup:/cgroup-in:ro -v $(pwd)/jdk20-jdk:/opt/jdk:z fedora:36 /opt/jdk/bin/java -Xlog:os+container=trace -version 2>&1 | grep 'Memory Limit'
[0.001s][trace][os,container] Memory Limit is: -2
[0.055s][trace][os,container] Memory Limit is: -2
```

Expected:

```
[0.001s][trace][os,container] Memory Limit is: 314572800
[0.068s][trace][os,container] Memory Limit is: 314572800
```

File: test/hotspot/jtreg/containers/docker/DockerBasicTest.java
Comment (casparcwang) [reply test/hotspot/jtreg/containers/docker/DockerBasicTest.java]: done

File: test/hotspot/jtreg/containers/docker/DockerBasicTest.java
Comment (casparcwang) [reply test/hotspot/jtreg/containers/docker/DockerBasicTest.java]: done

**Comment by casparcwang** [comment]
> We definitely need regression tests for memory/cpu limits with those additional cgroup mounts. Please take a look at `test/hotspot/jtreg/containers/docker/TestCPUAwareness.java` and `test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java`
> 
> For cgroups v2 the issues isn't fixed. The paths for cg v2 are being set here: https://github.com/openjdk/jdk/blob/98da03af50e2372817a7b5e381eea5ee6f2cb919/src/hotspot/os/linux/cgroupSubsystem_linux.cpp#L303..L313.
> 
> Example of running (with your patch on a cg v2 system):
> 
> ```
> $ sudo podman run --rm -ti --memory=300M --memory-swap=300M -v /sys/fs/cgroup:/cgroup-in:ro -v $(pwd)/jdk20-jdk:/opt/jdk:z fedora:36 /opt/jdk/bin/java -Xlog:os+container=trace -version 2>&1 | grep 'Memory Limit'
> [0.001s][trace][os,container] Memory Limit is: -2
> [0.055s][trace][os,container] Memory Limit is: -2
> ```
> 
> Expected:
> 
> ```
> [0.001s][trace][os,container] Memory Limit is: 314572800
> [0.068s][trace][os,container] Memory Limit is: 314572800
> ```


I have add a cgroup v2 test with duplicate mount info, now it's working for cgroup v2.

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (jerboaa) [inline src/hotspot/os/linux/cgroupSubsystem_linux.cpp:346]: Couldn't we simplify this? For example by changing:

```
      if (!cgroupv2_mount_point_found && strcmp("cgroup2", tmp_fs_type) == 0) {
        cgroupv2_mount_point_found = true;
        any_cgroup_mounts_found = true;
        for (int i = 0; i < CG_INFO_LENGTH; i++) {
          assert(cg_infos[i]._mount_path == NULL, "_mount_path memory stomping");
          cg_infos[i]._mount_path = os::strdup(tmp_mount_point);
        }
```

to something like this:

```
      if (strcmp("cgroup2", tmp_fs_type) == 0) {
        cgroupv2_mount_point_found = true;
        any_cgroup_mounts_found = true;
        for (int i = 0; i < CG_INFO_LENGTH; i++) {
          set_controller_paths(cg_infos, i, "(cg2, unified)", tmpmount, tmproot);
        }
```

It would entail modifying the `sscanf` to something like `sscanf(p, "%*d %*d %*d:%*d %s %s %*[^-]- %s %*s %*s", tmproot, tmpmount, tmp_fs_type) == 3`. We could also get rid of duplicate local var `tmp_mount_point`.

File: test/hotspot/jtreg/containers/docker/TestCPUAwareness.java
Comment (jerboaa) [inline test/hotspot/jtreg/containers/docker/TestCPUAwareness.java:80]: It should be sufficient to test the `true` case once. I.e. one test `testCpuQuotaAndPeriod(50*1000, 100*1000, true);` should be enough (as this would extend container test runs needlessly).

File: test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java
Comment (jerboaa) [inline test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java:73]: One test with the additional cgroup mounts should be sufficient. I suggest:

```
testMemoryLimit("500m", "524288000", true /* additional cgroup mount */);
```

**Review by jerboaa — COMMENTED** [review]
Thanks for the updates! A few more thoughts.

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (casparcwang) [reply src/hotspot/os/linux/cgroupSubsystem_linux.cpp]: done

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (casparcwang) [reply src/hotspot/os/linux/cgroupSubsystem_linux.cpp]: done

File: test/hotspot/jtreg/containers/docker/TestCPUAwareness.java
Comment (casparcwang) [reply test/hotspot/jtreg/containers/docker/TestCPUAwareness.java]: done

File: test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java
Comment (casparcwang) [reply test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java]: done

**Review by jerboaa — APPROVED** [review]
Looks good to me. Feel free to add the suggestion.

File: src/hotspot/os/linux/cgroupSubsystem_linux.cpp
Comment (casparcwang) [reply src/hotspot/os/linux/cgroupSubsystem_linux.cpp:304]: Done. Thank you for your patience review.

**Review by jerboaa — APPROVED** [review]

**Comment by casparcwang** [comment]
Is it OK to be pushed? Or shall we wait for one more review?

**Comment by jerboaa** [comment]
> Is it OK to be pushed? Or shall we wait for one more review?

A second review would be good. Perhaps @iklam has some cycles?

File: test/hotspot/jtreg/containers/docker/DockerBasicTest.java
Comment (iklam) [inline test/hotspot/jtreg/containers/docker/DockerBasicTest.java:94]: I think it's worth commenting what you are testing for here. Maybe something like

```
Duplicated cgroup mounts should be handled by the container detection
code and should not cause any error/warning output.
```

The `[os,container]` may not match exactly. Some extra spaces may be inserted before the close bracket if longer log tags have been printed. It should be changed to 

```
shouldNotMatch("[os,container *]");
```

(Please run the test on a build without your fix to verified that it would indeed fail).

**Review by iklam — APPROVED** [review]
Looks good to me. Just a small nit on the test,

File: test/hotspot/jtreg/containers/docker/DockerBasicTest.java
Comment (casparcwang) [reply test/hotspot/jtreg/containers/docker/DockerBasicTest.java:96]: Thank you for the review suggestions.

The comment has been added. When turn on `-Xlog:os+container=trace `, the output is like the following
```
[0.000s][trace][os,container] OSContainer::init: Initializing Container Support
[0.000s][debug][os,container] Detected optional pids controller entry in /proc/cgroups
[0.000s][debug][os,container] Duplicate pids controllers detected. Picking /sys/fs/cgroup/pids, skipping /cgroups-in/pids.
[0.000s][debug][os,container] Duplicate cpuset controllers detected. Picking /sys/fs/cgroup/cpuset, skipping /cgroups-in/cpuset.
[0.000s][debug][os,container] Duplicate cpu controllers detected. Picking /sys/fs/cgroup/cpu,cpuacct, skipping /cgroups-in/cpu,cpuacct.
[0.001s][debug][os,container] Duplicate cpuacct controllers detected. Picking /sys/fs/cgroup/cpu,cpuacct, skipping /cgroups-in/cpu,cpuacct.
[0.001s][debug][os,container] Duplicate memory controllers detected. Picking /sys/fs/cgroup/memory, skipping /cgroups-in/memory.
[0.001s][debug][os,container] Detected cgroups hybrid or legacy hierarchy, using cgroups v1 controllers
[0.001s][trace][os,container] Path to /memory.use_hierarchy is /sys/fs/cgroup/memory/memory.use_hierarchy
[0.001s][trace][os,container] Use Hierarchy is: 1
[0.001s][trace][os,container] Path to /cpu.cfs_quota_us is /sys/fs/cgroup/cpu,cpuacct/cpu.cfs_quota_us
[0.001s][trace][os,container] CPU Quota is: -1
[0.001s][trace][os,container] Path to /cpu.cfs_period_us is /sys/fs/cgroup/cpu,cpuacct/cpu.cfs_period_us
[0.001s][trace][os,container] CPU Period is: 100000
[0.001s][trace][os,container] OSContainer::active_processor_count: 16
[0.001s][trace][os,container] CgroupSubsystem::active_processor_count (cached): 16
[0.001s][trace][os,container] total physical memory: 33396502528
[0.001s][trace][os,container] Path to /memory.limit_in_bytes is /sys/fs/cgroup/memory/memory.limit_in_bytes
[0.001s][trace][os,container] Memory Limit is: 9223372036854771712
[0.001s][trace][os,container] Non-Hierarchical Memory Limit is: Unlimited
[0.001s][trace][os,container] Path to /memory.stat is /sys/fs/cgroup/memory/memory.stat
[0.001s][trace][os,container] Hierarchical Memory Limit is: 9223372036854771712
[0.001s][trace][os,container] Hierarchical Memory Limit is: Unlimited
[0.001s][debug][os,container] container memory limit unlimited: -1, using host value 33396502528
[0.003s][trace][os,container] CgroupSubsystem::active_processor_count (cached): 16
[0.045s][trace][os,container] Path to /cpu.cfs_quota_us is /sys/fs/cgroup/cpu,cpuacct/cpu.cfs_quota_us
[0.045s][trace][os,container] CPU Quota is: -1
[0.045s][trace][os,container] Path to /cpu.cfs_period_us is /sys/fs/cgroup/cpu,cpuacct/cpu.cfs_period_us
[0.045s][trace][os,container] CPU Period is: 100000
[0.045s][trace][os,container] OSContainer::active_processor_count: 16
[0.052s][trace][os,container] total physical memory: 33396502528
[0.052s][trace][os,container] Path to /memory.limit_in_bytes is /sys/fs/cgroup/memory/memory.limit_in_bytes
[0.052s][trace][os,container] Memory Limit is: 9223372036854771712
[0.052s][trace][os,container] Non-Hierarchical Memory Limit is: Unlimited
[0.052s][trace][os,container] Path to /memory.stat is /sys/fs/cgroup/memory/memory.stat
[0.052s][trace][os,container] Hierarchical Memory Limit is: 9223372036854771712
[0.052s][trace][os,container] Hierarchical Memory Limit is: Unlimited
[0.052s][debug][os,container] container memory limit unlimited: -1, using host value 33396502528
```

so `"[os,container *]"` will not match the output, and add an extra `*` will make failed test pass.

Comment (iklam) [reply test/hotspot/jtreg/containers/docker/DockerBasicTest.java:96]: I am sorry, I meant to use a regular expression to account for any possible space characters that comes after the `os,container`. Example:

```
$ java -Xlog:all -Xlog:os+container=debug --version | grep -i container | tail -2
[0.033s][debug][os,container   ] container memory limit unlimited: -1, using host value
[0.033s][debug][os,container   ] container memory limit unlimited: -1, using host value
```

Instead of `shouldNotContain`, which takes a plain string, you should use `shouldNotMatch`, which takes a regular expression:

```
.shouldNotMatch("\[os,container *\]");
```

Comment (casparcwang) [reply test/hotspot/jtreg/containers/docker/DockerBasicTest.java]: Thanks for the suggestion, I have changed it to regular expression `.shouldNotMatch("\\[os,container *\\]");`

**Review by iklam — APPROVED** [review]

**Review by jerboaa — APPROVED** [review]
Still good for me.

**Comment by jerboaa** [comment]
@casparcwang Feel free to issue `/integrate` again and I'll sponsor. Thanks!

**Comment by casparcwang** [comment]
> @casparcwang Feel free to issue `/integrate` again and I'll sponsor. Thanks!

Thank you very much!

**Pull request closed without merging**
